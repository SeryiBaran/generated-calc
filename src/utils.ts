type Sign = "+" | "-" | "/" | "*";
type Signs = Sign[];

export function generateCalcCode(
  minNumber: number,
  maxNumber: number,
  signs: Signs
) {
  let calcCode = "";

  calcCode += `
// Generated calc program

#include <iostream>
using namespace std;

int main() {

  string operation;
  int a;
  int b;

  cout << "Enter an operation: ";
  cin >> operation;

  cout << "Enter an a: ";
  cin >> a;

  cout << "Enter an b: ";
  cin >> b;
`;

  signs.forEach((sign) => {
    for (let i = minNumber; i < maxNumber + 1; i++) {
      for (let j = minNumber; j < maxNumber + 1; j++) {
        calcCode += `
  ${
    i === minNumber && j === minNumber ? "" : "else "
  }if (a == ${i} && operation == "${sign}" && b == ${j}) { ${
        sign === "/" && (i === 0 || j === 0)
          ? 'cout << "Zero division! Exit!" << endl; return 1;'
          : `cout << ${i} ${sign} ${j} << endl;`
      } }`;
      }
    }
  });

  calcCode += `
  else if (${signs.map((sign) => `operation != "${sign}"`).join(" && ")}) {
    cout << "Unknown operation (allowed - '${signs.join("', '")}')" << endl;
    return 1;
  }
  else if (a < ${minNumber} || a > ${maxNumber} || b < ${minNumber} || b > ${maxNumber}) {
    cout << "Out of range (Min - ${minNumber}, max - ${maxNumber})" << endl;
    return 1;
  }
  return 0;
}
`;

  return calcCode;
}
