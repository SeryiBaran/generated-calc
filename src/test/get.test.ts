import { expect, test } from "vitest";
import { generateCalcCode } from "../utils";

test("Generate code", () => {
  const result = generateCalcCode(0, 4, ["+", "-", "/", "*"]);
  expect(result).toMatchSnapshot();
});
