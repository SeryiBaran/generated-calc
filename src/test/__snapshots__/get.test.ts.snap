// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generate code 1`] = `
"
// Generated calc program

#include <iostream>
using namespace std;

int main() {

  string operation;
  int a;
  int b;

  cout << \\"Enter an operation: \\";
  cin >> operation;

  cout << \\"Enter an a: \\";
  cin >> a;

  cout << \\"Enter an b: \\";
  cin >> b;

  if (a == 0 && operation == \\"+\\" && b == 0) { cout << 0 + 0 << endl; }
  else if (a == 0 && operation == \\"+\\" && b == 1) { cout << 0 + 1 << endl; }
  else if (a == 0 && operation == \\"+\\" && b == 2) { cout << 0 + 2 << endl; }
  else if (a == 0 && operation == \\"+\\" && b == 3) { cout << 0 + 3 << endl; }
  else if (a == 0 && operation == \\"+\\" && b == 4) { cout << 0 + 4 << endl; }
  else if (a == 1 && operation == \\"+\\" && b == 0) { cout << 1 + 0 << endl; }
  else if (a == 1 && operation == \\"+\\" && b == 1) { cout << 1 + 1 << endl; }
  else if (a == 1 && operation == \\"+\\" && b == 2) { cout << 1 + 2 << endl; }
  else if (a == 1 && operation == \\"+\\" && b == 3) { cout << 1 + 3 << endl; }
  else if (a == 1 && operation == \\"+\\" && b == 4) { cout << 1 + 4 << endl; }
  else if (a == 2 && operation == \\"+\\" && b == 0) { cout << 2 + 0 << endl; }
  else if (a == 2 && operation == \\"+\\" && b == 1) { cout << 2 + 1 << endl; }
  else if (a == 2 && operation == \\"+\\" && b == 2) { cout << 2 + 2 << endl; }
  else if (a == 2 && operation == \\"+\\" && b == 3) { cout << 2 + 3 << endl; }
  else if (a == 2 && operation == \\"+\\" && b == 4) { cout << 2 + 4 << endl; }
  else if (a == 3 && operation == \\"+\\" && b == 0) { cout << 3 + 0 << endl; }
  else if (a == 3 && operation == \\"+\\" && b == 1) { cout << 3 + 1 << endl; }
  else if (a == 3 && operation == \\"+\\" && b == 2) { cout << 3 + 2 << endl; }
  else if (a == 3 && operation == \\"+\\" && b == 3) { cout << 3 + 3 << endl; }
  else if (a == 3 && operation == \\"+\\" && b == 4) { cout << 3 + 4 << endl; }
  else if (a == 4 && operation == \\"+\\" && b == 0) { cout << 4 + 0 << endl; }
  else if (a == 4 && operation == \\"+\\" && b == 1) { cout << 4 + 1 << endl; }
  else if (a == 4 && operation == \\"+\\" && b == 2) { cout << 4 + 2 << endl; }
  else if (a == 4 && operation == \\"+\\" && b == 3) { cout << 4 + 3 << endl; }
  else if (a == 4 && operation == \\"+\\" && b == 4) { cout << 4 + 4 << endl; }
  if (a == 0 && operation == \\"-\\" && b == 0) { cout << 0 - 0 << endl; }
  else if (a == 0 && operation == \\"-\\" && b == 1) { cout << 0 - 1 << endl; }
  else if (a == 0 && operation == \\"-\\" && b == 2) { cout << 0 - 2 << endl; }
  else if (a == 0 && operation == \\"-\\" && b == 3) { cout << 0 - 3 << endl; }
  else if (a == 0 && operation == \\"-\\" && b == 4) { cout << 0 - 4 << endl; }
  else if (a == 1 && operation == \\"-\\" && b == 0) { cout << 1 - 0 << endl; }
  else if (a == 1 && operation == \\"-\\" && b == 1) { cout << 1 - 1 << endl; }
  else if (a == 1 && operation == \\"-\\" && b == 2) { cout << 1 - 2 << endl; }
  else if (a == 1 && operation == \\"-\\" && b == 3) { cout << 1 - 3 << endl; }
  else if (a == 1 && operation == \\"-\\" && b == 4) { cout << 1 - 4 << endl; }
  else if (a == 2 && operation == \\"-\\" && b == 0) { cout << 2 - 0 << endl; }
  else if (a == 2 && operation == \\"-\\" && b == 1) { cout << 2 - 1 << endl; }
  else if (a == 2 && operation == \\"-\\" && b == 2) { cout << 2 - 2 << endl; }
  else if (a == 2 && operation == \\"-\\" && b == 3) { cout << 2 - 3 << endl; }
  else if (a == 2 && operation == \\"-\\" && b == 4) { cout << 2 - 4 << endl; }
  else if (a == 3 && operation == \\"-\\" && b == 0) { cout << 3 - 0 << endl; }
  else if (a == 3 && operation == \\"-\\" && b == 1) { cout << 3 - 1 << endl; }
  else if (a == 3 && operation == \\"-\\" && b == 2) { cout << 3 - 2 << endl; }
  else if (a == 3 && operation == \\"-\\" && b == 3) { cout << 3 - 3 << endl; }
  else if (a == 3 && operation == \\"-\\" && b == 4) { cout << 3 - 4 << endl; }
  else if (a == 4 && operation == \\"-\\" && b == 0) { cout << 4 - 0 << endl; }
  else if (a == 4 && operation == \\"-\\" && b == 1) { cout << 4 - 1 << endl; }
  else if (a == 4 && operation == \\"-\\" && b == 2) { cout << 4 - 2 << endl; }
  else if (a == 4 && operation == \\"-\\" && b == 3) { cout << 4 - 3 << endl; }
  else if (a == 4 && operation == \\"-\\" && b == 4) { cout << 4 - 4 << endl; }
  if (a == 0 && operation == \\"/\\" && b == 0) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 0 && operation == \\"/\\" && b == 1) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 0 && operation == \\"/\\" && b == 2) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 0 && operation == \\"/\\" && b == 3) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 0 && operation == \\"/\\" && b == 4) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 1 && operation == \\"/\\" && b == 0) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 1 && operation == \\"/\\" && b == 1) { cout << 1 / 1 << endl; }
  else if (a == 1 && operation == \\"/\\" && b == 2) { cout << 1 / 2 << endl; }
  else if (a == 1 && operation == \\"/\\" && b == 3) { cout << 1 / 3 << endl; }
  else if (a == 1 && operation == \\"/\\" && b == 4) { cout << 1 / 4 << endl; }
  else if (a == 2 && operation == \\"/\\" && b == 0) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 2 && operation == \\"/\\" && b == 1) { cout << 2 / 1 << endl; }
  else if (a == 2 && operation == \\"/\\" && b == 2) { cout << 2 / 2 << endl; }
  else if (a == 2 && operation == \\"/\\" && b == 3) { cout << 2 / 3 << endl; }
  else if (a == 2 && operation == \\"/\\" && b == 4) { cout << 2 / 4 << endl; }
  else if (a == 3 && operation == \\"/\\" && b == 0) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 3 && operation == \\"/\\" && b == 1) { cout << 3 / 1 << endl; }
  else if (a == 3 && operation == \\"/\\" && b == 2) { cout << 3 / 2 << endl; }
  else if (a == 3 && operation == \\"/\\" && b == 3) { cout << 3 / 3 << endl; }
  else if (a == 3 && operation == \\"/\\" && b == 4) { cout << 3 / 4 << endl; }
  else if (a == 4 && operation == \\"/\\" && b == 0) { cout << \\"Zero division! Exit!\\" << endl; return 1; }
  else if (a == 4 && operation == \\"/\\" && b == 1) { cout << 4 / 1 << endl; }
  else if (a == 4 && operation == \\"/\\" && b == 2) { cout << 4 / 2 << endl; }
  else if (a == 4 && operation == \\"/\\" && b == 3) { cout << 4 / 3 << endl; }
  else if (a == 4 && operation == \\"/\\" && b == 4) { cout << 4 / 4 << endl; }
  if (a == 0 && operation == \\"*\\" && b == 0) { cout << 0 * 0 << endl; }
  else if (a == 0 && operation == \\"*\\" && b == 1) { cout << 0 * 1 << endl; }
  else if (a == 0 && operation == \\"*\\" && b == 2) { cout << 0 * 2 << endl; }
  else if (a == 0 && operation == \\"*\\" && b == 3) { cout << 0 * 3 << endl; }
  else if (a == 0 && operation == \\"*\\" && b == 4) { cout << 0 * 4 << endl; }
  else if (a == 1 && operation == \\"*\\" && b == 0) { cout << 1 * 0 << endl; }
  else if (a == 1 && operation == \\"*\\" && b == 1) { cout << 1 * 1 << endl; }
  else if (a == 1 && operation == \\"*\\" && b == 2) { cout << 1 * 2 << endl; }
  else if (a == 1 && operation == \\"*\\" && b == 3) { cout << 1 * 3 << endl; }
  else if (a == 1 && operation == \\"*\\" && b == 4) { cout << 1 * 4 << endl; }
  else if (a == 2 && operation == \\"*\\" && b == 0) { cout << 2 * 0 << endl; }
  else if (a == 2 && operation == \\"*\\" && b == 1) { cout << 2 * 1 << endl; }
  else if (a == 2 && operation == \\"*\\" && b == 2) { cout << 2 * 2 << endl; }
  else if (a == 2 && operation == \\"*\\" && b == 3) { cout << 2 * 3 << endl; }
  else if (a == 2 && operation == \\"*\\" && b == 4) { cout << 2 * 4 << endl; }
  else if (a == 3 && operation == \\"*\\" && b == 0) { cout << 3 * 0 << endl; }
  else if (a == 3 && operation == \\"*\\" && b == 1) { cout << 3 * 1 << endl; }
  else if (a == 3 && operation == \\"*\\" && b == 2) { cout << 3 * 2 << endl; }
  else if (a == 3 && operation == \\"*\\" && b == 3) { cout << 3 * 3 << endl; }
  else if (a == 3 && operation == \\"*\\" && b == 4) { cout << 3 * 4 << endl; }
  else if (a == 4 && operation == \\"*\\" && b == 0) { cout << 4 * 0 << endl; }
  else if (a == 4 && operation == \\"*\\" && b == 1) { cout << 4 * 1 << endl; }
  else if (a == 4 && operation == \\"*\\" && b == 2) { cout << 4 * 2 << endl; }
  else if (a == 4 && operation == \\"*\\" && b == 3) { cout << 4 * 3 << endl; }
  else if (a == 4 && operation == \\"*\\" && b == 4) { cout << 4 * 4 << endl; }
  else if (operation != \\"+\\" && operation != \\"-\\" && operation != \\"/\\" && operation != \\"*\\") {
    cout << \\"Unknown operation (allowed - '+', '-', '/', '*')\\" << endl;
    return 1;
  }
  else if (a < 0 || a > 4 || b < 0 || b > 4) {
    cout << \\"Out of range (Min - 0, max - 4)\\" << endl;
    return 1;
  }
  return 0;
}
"
`;
